core/src/main/java/io/grpc/internal/ServerImpl.java

```diff
@@ -388,6 +409,7 @@ public ServerTransportListener transportCreated(ServerTransport transport) {
     public void serverShutdown() {
       ArrayList<ServerTransport> copiedTransports;
       Status shutdownNowStatusCopy;
+      awaitStartComplete();
```

Comment:
This looks dangerous.

I'd much rather we set `activeTransportServers = transportServers.size()` within the initial synchronization block of start() to prevent needing to block here.

Reply:
set `activeTransportServers = transportServers.size()` is broken if one of `ts.start()` throws because not enough times of `listern.serverShutdown()` will be called .

Reply:
Seems we need redesign multiple transportServers support in future. It's just too inconvenient now. For now the fix seems working. Can you elaborate what is the danger?

Reply:
Now I see the danger, because `serverShutdown()` is running in the event loop, if another `ts.start()` submits something into the event loop while `awaitStartComplete()` in  `serverShutdown()`, then it will be deadlock again.